### PR TITLE
File Browser Filter

### PR DESCRIPTION
**Pull request by echarles**
## References

This PR adresses https://github.com/jupyterlab/jupyterlab/issues/8157.

For now, only the files of the current folder are considered, see also the TODO section hereafter.

## Code changes

- Changes in `filebrowser` and `filebrowser-extension` TS packages.
- Use `FilterFileBrowserModel` instead of `FileBrowserModel` and create a `search.ts` for input text

## User-facing changes

filefilter

## Backwards-incompatible changes

None.

## TODO

- [ ] Fix CSS issue when file listing is larger than the display layout.
- [ ] Consider to filter on folders (the existing `FilterFileBrowserModel` keeps the folders.
- [ ] Consider to filter on subfolders if this feature is required.
- [ ] Consider highlighting in the filenames the given filter.

### CODE REVIEW

**Comment by telamonian** [comment]
@echarles Looks good so far. I think the UI needs some work.

- insert the search bar above the breadcrumbs
  - the search bar probably shouldn't cover up the breadcrumbs, because then there's no info displayed to the user about what their cwd is. Instead, insert the search UI above the breadcrumb

- get rid of the search toolbar button
  - at the same time, I *hate* list UI's that shift large elements up and down depending on state. Having things jump up and down makes me seasick. I think most users would appreciate the existence of a search bar, so just get rid of the toolbar button and have it always visible
  - the issue could also be mitigated by animating the transition, but that's much fancier than our usual approach to styling

- Change the placeholder text to `filter files by name`
  - intuitively, I think most users would expect a "file search" functionality to recurse into dirs, or possibly search their whole tree starting from root
  - since the current proposal is for a shallow search, it would be clearer to call this a "file filter"

**Comment by echarles** [comment]
Thx @telamonian for your review and feedbacks. I will push some changes to take them.

> at the same time, I hate list UI's that shift large elements up and down depending on state. Having things jump up and down makes me seasick. I think most users would appreciate the existence of a search bar, so just get rid of the toolbar button and have it always visible

I would like just to think a bit more to having the input search box always visible instead of activating it via the toolbar and replacing the breadcrumbs. With the `always visible` option, we loose some space on the file browser area. I am fine for that, but we may need to think a bit more and maybe have other opinions and options.

cc/ @tgeorgeux

**Comment by echarles** [comment]
I have updated the placeholder which is now `Filter files by name` as @telamonian has suggested.

For the search box UI, I have also implemented @telamonian suggestion. I will put this topic at today weekly meeting to gather broader community feedback as this change is quite visible for the user. I see 3 options:

1. Toogle via toolbar (replace breadcrumbs)

Screenshot 2020-07-01 at 15 25 52

2. Always visible

Screenshot 2020-07-01 at 15 30 31

3. Side-panel (a-la-vscode)

Screenshot 2020-07-01 at 15 36 40

You can also mix them, e.g. 1. Toogle via toolbar + keep the breadcrumbs. Side-panel (a-la-vscode) would have more impact in terms of implementation, and also makes less sense for now as the goal of this PR is restricted to filter on files in the current directory, so not a real search on file names in the nested folders.

**Comment by tgeorgeux** [comment]
Option 2 follows the design patterns the best at this point. I would make a few small adjustments:
- Add 4px padding all around the search box.
- Switch out the Icon for the [JupyterLab Search Icon ](https://github.com/jupyterlab/jupyterlab/blob/master/packages/ui-components/style/icons/toolbar/search.svg)(we are using material Icons, that one looks like it's from the Ionic Icon set).

**Comment by echarles** [comment]
@tgeorgeux padding and icon updated.

@telamonian Tim has asked to use mui icon, so I have update the InputGroup rightIcon to use LabIcon (and impact the ones using that) - Can you have a look at that and provide any feedback (eg. I use `rightIcon="ui-components:search"`, not sure if you would prefer `rightIcon="search"`

Screenshot 2020-07-01 at 17 48 21

The impact of this is that eg. extension manager now also uses the LabIcon, which was not the case before (Discussed about that with Tim and he said he would prefer to align all search icons anyway).

Screenshot 2020-07-01 at 17 48 33

**Comment by echarles** [comment]
@telamonian I think you said filter *above* the breadcrumbs, so here it is

Screenshot 2020-07-01 at 18 07 42

**Comment by blink1073** [comment]
Command palette uses [`matchSumOfDeltas`](https://github.com/jupyterlab/lumino/blob/52b15d88cd0e647c40bd1d6a265817b7cc6458a2/packages/algorithm/src/string.ts#L132) to do its [filtering](https://github.com/jupyterlab/lumino/blob/7831eb2725f291a6c99fc72ae0b9be42afe0ca8c/packages/widgets/src/commandpalette.ts#L1246).

**Comment by echarles** [comment]
Thx @blink1073. Very useful links that helped me to go fuzzy (with a small optional addition attributed on the IContent model). This is ready for review.

filefilter

File: packages/filebrowser/src/search.tsx [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      rightIcon="ui-components:search"
+      placeholder={props.placeholder}
+      onChange={handleChange}
+      value={filter}
+    />
+  );
+};
+
+/**
+ * A widget which hosts a ...
```
Comment (blink1073) [inline packages/filebrowser/src/search.tsx:131]: Missing an ending ;)

**Comment by blink1073** [comment]
Let's make sure this still passes after merging #8666

**Comment by blink1073** [comment]
Kicking CI, I'll cut an alpha release after we merge this.

**Comment by blink1073** [comment]
Looks like the filebrowser example needs updating: 

```
##[error]src/index.ts(97,5): error TS2739: Type 'FileBrowserModel' is missing the following properties from type 'FilterFileBrowserModel': setFilter, _filter
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by echarles** [comment]
@blink1073 Example is fixed and CI has been green

@telamonian I have added a user setting `useFuzzyFilter` which is true by default - Set it to false for normal filter

@tgeorgeux The mapped string are now shown like in the command sidebar (black bold)

**Comment by echarles** [comment]
ezgif-1-23da821276c0

**Comment by blink1073** [comment]
Testing the new auto-cancel bot for GH Actions, opening and closing a couple times.

**Comment by blink1073** [comment]
Success!

**Comment by blink1073** [comment]
Failure in js-filebrowser:

```
 FAIL  test/openfiledialog.spec.ts
  ● @jupyterlab/filebrowser › FileDialog.getExistingDirectory() › should return one selected directory

    expect(received).toBe(expected) // Object.is equality

    Expected: "directory"
    Received: "file"
```

**Comment by echarles** [comment]
@blink1073 My last commit had unexpected side effect which I have fixed now. Thx for spotting this

CI is green (except Linux Tests / Linux (docs, 3.8) which fails after 78m, unrelated to this PR).

**Comment by blink1073** [comment]
Nice work @echarles!

**Comment by michaelaye** [comment]
Is there a work-around to enable filtering on folders in JLab 3.0 ?

**Comment by jpcaram** [comment]
I would like to filter folders too! Is there a setting?

**Comment by vidartf** [comment]
> I would like to filter folders too! Is there a setting?

AFAIK, there is no such setting. You should probably make a new issue with this feature request if you want it 👍

**Pull request merged**